Fix #10399 - Long elapsed times are not recorded correctly

jmarrec opened the pull request:
Pull request overview
---------------------
 - Fixes #10399

I implemented a Timer class, that I made re-startable (even though not needed right now) like a stopwatch, in case we do want to use that for timings specific bits of the code.


### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

jmarrec committed:
Initial Timer class

jmarrec committed:
un-auto return types and make it restartable

jmarrec committed:
Unprivate some members to conform to E+ conventions + add testing for it

Myoldmopar commented on src/EnergyPlus/CMakeLists.txt:
```diff
@@ -602,6 +602,8 @@ set(SRC
     TarcogShading.cc
     TarcogShading.hh
+    Timer.cc
+    Timer.hh
```
Hey I've seen that before!  

https://github.com/NREL/EnergyPlusArchive/blob/master/SourceCode/Timer.h

Myoldmopar commented on tst/EnergyPlus/unit/Timer.unit.cc:
```diff
@@ -0,0 +1,120 @@
+        t.m_start =
+            Timer::ClockType::now() - (std::chrono::hours(2) + std::chrono::minutes(25) + std::chrono::seconds(51) + std::chrono::milliseconds(341));
+        t.tock();
+        EXPECT_EQ("02hr 25min 51.34sec", t.formatAsHourMinSecs());
```
Hmmm...  This feels...  Probably fine...

Myoldmopar reviewed (approved):
Yeah this is nice.  I'll do a quick build and check CI, but it's good here.

Myoldmopar commented:
Confirmed everything is :100: here, merging.  Thanks @jmarrec

jmarrec commented on src/EnergyPlus/Timer.cc:
```diff
@@ -0,0 +1,112 @@
+std::string Timer::formatAsHourMinSecs() const
+{
+    // We're working with ms as base.
+    // TODO: do we want to report elapsed time of 25 hours as 1 day 1hr 0min 0sec or 25hr 0min 0sec?
```
@Myoldmopar forgot to flag this
Do we want 1 days 1 hr xxx or 2hr

Myoldmopar replied:
I do not care...  Possibly 1 day 1 hr ??  I think either is fine though.  This would only matter in a corner case where `(runtime is **enormously** long) .and. (the user cant handle a 25 hour time stamp)` which should round down to zero percent of the EnergyPlus use cases.  Anyway, whatever is happening right now is fine.

jmarrec commented on src/EnergyPlus/Timer.cc:
```diff
@@ -0,0 +1,112 @@
+
+std::string Timer::formatAsHourMinSecs() const
+{
+    // We're working with ms as base.
+    // TODO: do we want to report elapsed time of 25 hours as 1 day 1hr 0min 0sec or 25hr 0min 0sec?
+    // constexpr auto num_ms_in_day = 1000 * 3600 * 24;
+    // if (duration().count() > num_ms_in_day) {
+    //     return fmt::format("{:.2%j days %Hhr %Mmin %Ssec}\n", duration());
+    // } else {
+    //     return fmt::format("{:%Hhr %Mmin %Ssec}\n", duration());
+    // }
```
@Myoldmopar  Assuming we want to keep 25 hr as an output format, just apply this:

```suggestion
    // We're working with ms as base.
```

jmarrec commented on tst/EnergyPlus/unit/Timer.unit.cc:
```diff
@@ -0,0 +1,120 @@
+    std::this_thread::sleep_for(std::chrono::milliseconds(62));
+    t.tock();
+    EXPECT_EQ(124, t.duration().count());
+    EXPECT_EQ(0.124, t.elapsedSeconds());
```
I just realized that Marvin actually reported 128. I'll adjust the test, we don't want it to fail when the system is busy.

It's really hidden to see the green check mark and just get an email. @Myoldmopar
* Do we have a test failure threshold or something? Is that warranted?
* We really need to bring back PR annotations

![image](https://github.com/NREL/EnergyPlus/assets/5479063/48ef56b1-5d7f-48c8-b13e-65dbc3d95e63)


https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-61cf2187918f76746b9eaa263db904f4aeffed56-x86_64-MacOS-10.18-clang-15.0.0.html

Myoldmopar replied:
For test failure threshold, we should have it set to fail if _any_ unit test fails.  If regressions fail, that will be fine, but we should still see the warnings.  I'm still upset at GH for just disabling commit comments on the PR page without an easy way to bring them back.  

IIRC, in my development branch of Decent CI, I operate on PRs directly where possible.  So it's easy to move the comments to directly on the PRs.  I'm open to pushing on that again, it's just a big distraction.

jmarrec committed:
Relax tolerance for timer test

cf https://github.com/NREL/EnergyPlus/pull/10529#discussion_r1620481489

jmarrec committed:
Relax more

Obviously if I were to sleep for like 10 seconds, the % difference would be much smaller. But this is a dumb test, and it's the fourth time I push (I squashed earlier)

Myoldmopar commented:
This is all happy, let's get it in to get today's merge queue moving.  Thanks for adding the robustness to the time test, @jmarrec
